remind.1
.TH REMIND 1 "6 April 1994"
.UC 4
.SH NAME
remind \- a sophisticated reminder service
.SH SYNOPSIS
.B remind [\fIoptions\fR] \fIfilename\fR [\fIdate\fR] [\fI*rep\fR] [\fItime\fR]
.SH DESCRIPTION
\fBRemind\fR reads the supplied \fIfilename\fR and executes the commands
found in it.  The commands are used to issue reminders and alarms.  Each
reminder or alarm can consist of a message sent to standard output, or
a program to be executed.
.PP
If \fIfilename\fR is specified as a single dash '-', then \fBRemind\fR
takes its input from standard input.  This also implicitly enables
the \fB\-o\fR option, described below.
.SH OPTIONS
.TP
.B \-n
The \fB\-n\fR option causes \fBRemind\fR to print the \fBnext\fR occurrence
of each reminder in a simple calendar format.  You can sort this by
date by piping the output through \fBsort(1)\fR.
.TP
.B \-r
The \fB\-r\fR option disables \fBRUN\fR directives and the \fBshell()\fR
function
.TP
.B \-c\fR\fIn\fR
The \fB\-c\fR option causes \fBRemind\fR to produce a calendar which is sent to
standard output.  If you supply a number \fIn\fR, then a calendar will
be generated for \fIn\fR months, starting with the current month.  By
default, a calendar for only the current month is produced.  If \fIn\fR
starts with '+', then a calendar for \fIn\fR weeks is produced.
.TP
.B \-w\fR\fIcol\fR[,\fIpad\fR[,\fIspc\fR]]]
The \fB\-w\fR option specifies the output width, padding and spacing
of the formatted calendar output.  \fICol\fR specifies the number of
columns in the output device, and defaults to 80.  \fIPad\fR specifies
how many lines to use to "pad" empty calendar boxes.  This defaults to
5.  If you have many reminders on certain days, which make your calendar
too large to fit on a page, you can try reducing \fIpad\fR to make the
empty boxes smaller.  \fISpc\fR specifies how many blank lines to leave
between the day number and the first reminder entry.  It defaults to 1.
.RS
.PP
Any of \fIcol\fR, \fIpad\fR or \fIspc\fR can be omitted, providing you
provide the correct number of commas.  Don't use any spaces in the option.
.RE
.TP
.B \-s\fR\fIn\fR
The \fB\-s\fR option is very similar to the \fB\-c\fR option, except
that the output calendar is not formatted.  It is listed in a "simple
format" which can be used as input for more sophisticated calendar-drawing
programs.  If \fIn\fR starts with "+", then it is interpreted as a number
of weeks.
.TP
.B \-p\fR\fIn\fR
The \fB\-p\fR option is very similar to the \fB\-s\fR option, except
that the output contains additional information for use by the
\fBrem2ps\fR program, which creates a PostScript calendar.  For this
option, \fIn\fR cannot start with "+"; it must specify a number of months.
.TP
.B \-m
The \fB\-m\fR option causes the \fB\-c\fR option to produce a calendar whose
first column is Monday rather than Sunday.  (This conforms to the international
standard.)  It does \fInot\fR, however, affect the \fB\-p\fR or \fB\-s\fR
options.
.TP
.B \-v
The \fB\-v\fR option makes the output of \fBRemind\fR slightly more verbose.
.TP
.B \-o
The \fB\-o\fR option causes \fBRemind\fR to ignore all \fBONCE\fR directives.
.TP
.B \-t
The \fB\-t\fR option causes \fBRemind\fR to trigger all non-expired reminders,
regardless of the \fIdelta\fR supplied for each reminder.
.TP
.B \-h
The \fB\-h\fR option ("hush...") suppresses certain warning and information
messages.  In particular, if no reminders are triggered, this mode
produces no output.
.TP
\fB\-a\fR (UNIX and OS/2 only)
The \fB\-a\fR option causes \fBRemind\fR not to immediately trigger timed
reminders which would also be queued.  It also causes \fBRemind\fR not to
place timed reminders in a calendar.
.TP
\fB\-q\fR (UNIX and OS/2 only)
The \fB\-q\fR option causes \fBRemind\fR not to queue timed reminders
for later execution.
.TP
\fB\-f\fR (UNIX and OS/2 only)
The \fB\-f\fR option causes \fBRemind\fR to remain in the foreground
when processing queued reminders, rather than forking off
a background process to handle them.
.TP
.B \-e
The \fB\-e\fR option diverts error messages (normally sent to the
standard error stream) to the standard output stream.
.TP
.B \-d\fR\fIchars\fR
The \fB-d\fR option enables certain debugging modes.  The \fIchars\fR
specify which modes to enable:
.RS 2
.TP
.B e
Echo all input lines
.TP
.B x
Trace all expression evaluation
.TP
.B t
Display all trigger date computation
.TP
.B v
Dump the variable table after execution of the reminder script
.TP
.B l
Echo lines when displaying error messages
.RE
.TP
\fB\-g\fR[\fBa|d\fR[\fBa|d\fR[\fBa|d\fR]]]
Normally, reminders are issued in the order in which they are encountered
in the reminder script.  The \fB\-g\fR option cause \fBRemind\fR to
sort reminders by date and time prior to issuing them.  The optional
\fBa\fR and \fBd\fR characters specify the sort order (ascending or
descending) for the date, time and priority fields.  See the section "Sorting
Reminders" for more information.
.TP
\fB\-b\fR[\fIn\fR]
Set the time format for the calendar and simple-calendar outputs.  \fIN\fR
can range from 0 to 2, with the default 0.  A value of 0 causes times
to be inserted in 12-hour (am/pm) format.  1 causes times to be inserted
in 24-hour format, and 2 inhibits the automatic insertion of times in the
calendar output.
.TP
\fB\-x\fR[\fIn\fR]
Sets the iteration limit for the \fBSATISFY\fR clause of a \fBREM\fR
command.  Defaults to 150.
.TP
\fB\-k\fR\fIcmd\fR
Instead of simply printing \fBMSG\fR-type
reminders, this causes them to be passed to the specific \fIcmd\fR.
You must use '%s' where you want the body to appear, and may need to
enclose this option in quotes.  Note that all shell characters in the
body of the reminder are escaped with a backslash, and the entire body
of the reminder is passed as a single argument.  Note that this option
\fBoverrides\fR the \fB\-r\fR option and the \fBRUN OFF\fR command.
.PP
.RS
As an example, suppose you have an X Window program called xmessage, which
pops up a window and displays its invocation arguments.  You could use:
.PP
.nf
		remind '-kxmessage %s &' ...
.fi
.PP
to have all of your \fBMSG\fR-type reminders processed using xmessage.
.PP
A word of warning:  It is very easy to spawn dozens of xmessage processes
with the above technique.  So be very careful.  Also, the \fIcmd\fR is passed
as an argument to \fBsprintf()\fR.  If you use formatting directives other
than %s, or use more than one %s directive, there's a good chance that
you'll crash \fBRemind\fR.  Finally, because all shell and whitespace
characters are escaped, the program you execute with the \fB\-k\fR
option must be prepared to handle the entire message as a single argument.
.RE
.TP
\fB\-z\fR[\fIn\fR] (UNIX and OS/2 only)
Runs \fBRemind\fR in the daemon mode.  If \fIn\fR is supplied, it
specifies how often (in minutes) \fBRemind\fR should wake up to
check if the reminder script has been changed.  \fIN\fR defaults
to 5, and can range from 5 to 60.  Note that the use of the
\fB\-z\fR option also enables the \fB\-f\fR option.
.TP
\fB\-u\fR\fIname\fR  (UNIX version only)
Runs \fBRemind\fR with the uid and gid of the user specified by \fIname\fR.
The option changes the uid and gid as described, and sets the
environment variables HOME, SHELL and USER to the home directory, shell,
and user name, respectively, of the specified user.  LOGNAME is also
set to the specified user name.  This option is meant for
use in shell scripts which mail reminders to all users.
.PP
.RS
Non-root users can also use the \fB\-u\fR option.  However, in this
case, it only changes the environment variables as described above.
It does not change the effective uid or gid.
.RE
.TP
\fB\-i\fR\fIvar\fR\fB=\fR\fIexpr\fR
Sets the value of the specified \fIvar\fR to \fIexpr\fR, and \fBpreserves\fR
\fIvar\fR.  \fIExpr\fR can be any valid \fBRemind\fR expression.  See the
section "Initializing Variables on the Command Line" for more details.
.PP
If you supply a \fIdate\fR on the command line, it must consist of
\fIday month year\fR, where \fIday\fR is the day of the month,
\fImonth\fR is at least the first three letters of the English name
of the month, and \fIyear\fR is a year (all 4 digits) from 1990 to
about 2075.  You can leave out the \fIday\fR, which then defaults to 1.
.PP
If you do supply a \fIdate\fR on the command line, then \fBRemind\fR uses
it, rather than the actual system date, as its notion of "today."
This lets you create calendars for future months, or test to see
how your reminders will be triggered in the future.  Similarly,
you can supply a \fItime\fR (in 24-hour format -- for example, 17:15) to
set \fBRemind\fR's notion of "now" to a particular time.  Supplying
a \fItime\fR on the command line also implicitly enables the \fB\-q\fR
option and disables the \fB\-z\fR option.
.PP
In addition, you can supply a \fIrepeat\fR parameter, which has the
form *\fInum\fR.  This causes \fBRemind\fR to be run \fInum\fR times,
with the date incrementing on each iteration.  You may have to enclose
the parameter in quotes to avoid shell expansion.  See the subsection
"Repeated Execution" in the section "Calendar Mode" for more
information.
.SH REMINDER FILES
.PP
\fBRemind\fR uses scripts to control its operation.  You can use any
text editor capable of creating plain ASCII files to create a
\fBRemind\fR script.  The commands inside a script can range from the
very simple and almost immediately understandable:
.PP
.nf
	REM 6 Jan MSG David's birthday
.fi
.PP
to the baroque and obscure:
.PP
.nf
	REM [trigger(date(thisyear, 1, 1) + 180)] ++5 OMIT \\
	sat sun BEFORE MSG [ord(thisyear-1980)] payment due %b!
.fi
.PP
A reminder file consists of commands, with one command per line.  Several
lines can be continued using the backslash character, as in the above
example.  In this case, all of the concatenated lines are treated as a
single line by \fBRemind\fR.  Note that if an error occurs, \fBRemind\fR
reports the line number of the last line of a continued line.
.PP
\fBRemind\fR ignores blank lines, and lines beginning with the '#' or ';'
characters.  You can use the semicolon as a comment character if you
wish to pass a \fBRemind\fR script through the C pre-processor, which
interprets the '#' character as the start of a pre-processing
directive.
.PP
\fBRemind\fR is not case sensitive; you can generally use any mixture of upper-
or lower-case for commands, parameters, invocation options, etc.
.SH THE REM COMMAND
.PP
The most powerful command in a \fBRemind\fR script is the \fBREM\fR command.
This command is responsible for issuing reminders.
Its syntax is:
.PP
.RS
\fBREM\fR [\fBONCE\fR] [\fIdate_spec\fR]
[\fIback\fR]
[\fIdelta\fR]
[\fIrepeat\fR]
[\fBPRIORITY\fR \fIprio\fR]
[\fBSKIP\fR | \fBBEFORE\fR | \fBAFTER\fR]
[\fBOMIT\fR \fIomit_list\fR]
[\fBAT\fR \fItime\fR [\fItdelta\fR] [\fItrepeat\fR]]
[\fBSCHED\fR \fIsched_function\fR]
[\fBUNTIL\fR \fIexpiry_date\fR]
[\fBSCANFROM\fR \fIscan_date\fR]
\fBMSG\fR | \fBMSF\fR | \fBRUN\fR | \fBCAL\fR | \fBSATISFY\fR |
\fBPS\fR | \fBPSFILE\fR
.I body
.RE
.PP
The parts of the \fBREM\fR command can be specified in any order, except
that the \fIbody\fR must come immediately after the \fBMSG\fR,
\fBRUN\fR, \fBCAL\fR, \fBPS\fR, \fBPSFILE\fR or \fBSATISFY\fR keyword.
.PP
The \fBREM\fR token is optional, providing that the remainder
of the command cannot be mistaken for another \fBRemind\fR command
such as \fBOMIT\fR or \fBRUN\fR.  The portion of the \fBREM\fR command
before the \fBMSG\fR, \fBMSF\fR \fBRUN\fR, \fBCAL\fR or \fBSATISFY\fR clause
is called a \fItrigger\fR.
.PP
.B "MSG, MSF, RUN, CAL, PS and PSFILE"
.PP
These keywords denote the \fItype\fR
of the reminder.  (\fBSATISFY\fR is more complicated and will be explained
later.)  A \fBMSG\fR-type reminder normally prints a message to the standard
output, after passing the \fIbody\fR through a special substitution filter,
described in the section "The Substitution Filter."  However, if you have
used the \fB\-k\fR command-line option, then \fBMSG\fR-type reminders are
passed to the appropriate program.  Note that the options \fB\-c\fR,
\fB\-s\fR, \fB\-p\fR and \fB\-n\fR disable the \fB\-k\fR option.
.PP
The \fBMSF\fR keyword is almost the same as the \fBMSG\fR keyword,
except that the reminder is formatted to fit into a paragraph-like
format.  Three system variables control the formatting of \fBMSF\fR-type
reminders - they are \fB$FirstIndent\fR, \fB$SubsIndent\fR and
\fB$FormWidth\fR.  They are discussed in the section "System Variables."
The \fBMSF\fR keyword causes the spacing of your reminder to be altered -
extra spaces are discarded, and two spaces are placed after periods and
other characters, as specified by the system variables \fB$EndSent\fR and
\fB$EndSentIg\fR.  Note that if the body of the reminder includes
newline characters (placed there with the %_ sequence), then the newlines
are treated as the beginnings of new paragraphs, and the \fB$FirstIndent\fR
indentation is used for the next line.  You can use two consecutive
newlines to have spaced paragraphs emitted from a single reminder body.
.PP
A \fBRUN\fR-type
reminder also passes the \fIbody\fR through the substitution filter, but
then executes the result as a system command.  A \fBCAL\fR-type reminder
is used only to place entries in the calendar produced when \fBRemind\fR
is run with the \fB\-c\fR, \fB\-s\fR or \fB\-p\fR options.
.PP
A \fBPS\fR or \fBPSFILE\fR-type reminder is used to pass PostScript code
directly to the printer when producing PostScript calendars.  This can
be used to shade certain calendar entries (see the psshade() function),
include graphics in the calendar,
or almost any other purpose you can think of.  You
should not use these types of reminders unless you are an expert PostScript
programmer.  The \fBPS\fR and \fBPSFILE\fR reminders are ignored unless
\fBRemind\fR is run with the \fB\-p\fR option.  See the section
"More about PostScript" for more details.
.PP
.B DATE SPECIFICATIONS
.PP
A \fIdate_spec\fR consists of zero to four parts.
These parts are
.I day
(day of month),
.I month
(month name),
.I year
and
.I weekday.
.I Month
and
.I weekday
are the English names of months and weekdays.  At least the first three
characters must be used.  The following are examples of the various parts of a
.I date_spec:
.TP
.I day:
1, 22, 31, 14, 3
.TP
.I month:
JANUARY, feb, March, ApR, may, Aug
.TP
.I year:
1990, 1993, 2030, 95 (interpreted as 1995).  The year can range
from 1990 to 2075.
.TP
.I weekday:
Monday, tue, Wed, THU, Friday, saturday, sundAy
.PP
Note that there can be several
.I weekday
components separated by spaces in a
.I date_spec.
.PP
.B INTERPRETATION OF DATE SPECIFICATIONS
.PP
The following examples show how date specifications are interpreted.
.PP
1. Null date specification - the reminder is triggered every day.
The trigger date for a specific run is simply the current system date.
.PP
2. Only
.I day
present.  The reminder is triggered on the specified day of each month.
The trigger date for a particular run is the closest such day to the
current system date.  For example:
.nf
	REM 1 MSG First of every month.
	REM 31 MSG 31st of every month that has 31 days.
.fi
.PP
3. Only
.I month
present.  The reminder is triggered every day of the specified month.
Example:
.nf
	REM Feb MSG Every day in February
.fi
.PP
4.
.I day
and
.I month
present.  Examples:
.nf
	REM 6 Jan MSG Every 6th of January
	REM Feb 29 MSG Every 29th of February
.fi
.PP
5.  Only
.I year
present. Example:
.nf
	REM 1991 MSG Every day in 1991
.fi
.PP
6.
.I year
and
.I day
present.  Examples:
.nf
	REM 1 1990 MSG 1st of every month in 1990
	REM 1992 23 MSG 23rd of every month in 1992
.fi
.PP
7.
.I year
and
.I month
present.  Examples:
.nf
	REM Feb 1991 MSG Every day in Feb 1991
	REM 1992 September MSG Every day in Sept 1992
.fi
.PP
8.
.I year, month
and
.I day
present.  Examples:
.nf
	REM 8 Jan 1991 MSG 8th January 1991.
	REM 1992 March 9 MSG 9th March 1992.
.fi
.PP
9.
.I weekday
only.  Examples:
.nf
	REM Sat MSG Every Saturday
	REM Mon Tue Wed Thu Fri MSG Every working day
	REM Monday Wednesday MSG Every Monday and Wednesday
.fi
.PP
10.
.I weekday
and
.I day
present.  Examples:
.nf
	REM Sat 1 MSG First Saturday of every month
	REM Mon Tue Wed Thu Fri 15 \\
		MSG 1st working day after 15th of every month
.fi
.PP
11.
.I weekday
and
.I month
present.  Examples:
.nf
	REM Mon March MSG Every Monday in March
	REM Mon Tue Wed Thu Fri Feb MSG Every working day in February
.fi
.PP
12.
.I weekday, month
and
.I day
present.  Examples:
.nf
	REM Mon 1 March MSG First Monday in March
	REM Sat Sun 15 July MSG First Sat or Sun on or after 15 July
.fi
.PP
13.
.I weekday
and
.I year
present.  Example:
.nf
	REM Sat Sun 1991 MSG Every Saturday and Sunday in 1991
.fi
.PP
14.
.I weekday, day
and
.I year
present.  Examples:
.nf
	REM Mon 15 1990 MSG 1st Mon after 15th of every month in 1990
	REM Mon Tue Wed Thu Fri 1 1990 \\
		MSG 1st working day of every month in 1990
.fi
.PP
15.
.I weekday, month
and
.I year
present.  Example:
.nf
	REM Mon Wed 1991 Feb MSG Every Mon and Wed in Feb 1991.
.fi
.PP
16.
.I weekday, day, month
and
.I year
present.  Example:
.nf
	REM Mon Tue Wed Thu Fri 28 Oct 1990 \\
		MSG 1st working day on or after 28 October 1990.
.fi
.PP
Note that when both
.I weekday
and
.I day
are specified,
.B Remind
chooses the first date on or after the specified
.I day
which also satisfies the
.I weekday
constraint.  It does this by picking the first date on or after the specified
.I day
which is listed in the list of
.I weekdays.
Thus, a reminder like:
.PP
.nf
	REM Mon Tue 28 Oct 1990 MSG Hi
.fi
.PP
would be issued only on Monday, 29 October, 1990.  It would not be issued
on Tuesday, 30 October, 1990, since the 29th is the first date to satisfy
the
.I weekday
constraints.
.PP
.B BACKWARD SCANNING
.PP
Sometimes, it is necessary to specify a date as being a set amount of
time before another date.  For example, the last Monday in a given
month is computed as the first Monday in the next month, minus 7 days.
The \fIback\fR specification in the reminder is used in this case:
.PP
.nf
	REM Mon 1 -7 MSG Last Monday of every month.
.fi
.PP
A \fIback\fR is specified with one or two dashes followed by an integer.
This causes \fBRemind\fR to move "backwards" from what would normally be the
trigger date.  The difference between \-\-7 and \-7 will be explained
when the \fBOMIT\fR keyword is described.
.PP
.B ADVANCE WARNING
.PP
For some reminders, it is appropriate to receive advance warning of the
event.  For example, you may wish to be reminded of someone's birthday
several days in advance.  The \fIdelta\fR portion of the \fBREM\fR command
achieves this.  It is specified as one or two "+" signs followed by a number
\fIn\fR.  Again, the difference between the "+" and "++" forms will
be explained under the \fBOMIT\fR keyword.
\fBRemind\fR will trigger the reminder on computed trigger date, as well as
on each of the \fIn\fR days before the event.  Here are some examples:
.PP
.nf
	REM 6 Jan +5 MSG Remind me of birthday 5 days in advance.
.fi
.PP
The above example would be triggered every 6th of January, as well as the
1st through 5th of January.
.PP
.B PERIODIC REMINDERS
.PP
We have already seen some built-in mechanisms for certain types of
periodic reminders.  For example, an event occurring every Wednesday
could be specified as:
.PP
.nf
	REM Wed MSG Event!
.fi
.PP
However, events which do not repeat daily, weekly, monthly or yearly require
another approach.  The \fIrepeat\fR component of the \fBREM\fR command
fills this need.  To use it, you must completely specify a date (year, month
and day, and optionally weekday.)  The \fIrepeat\fR component is an asterisk
followed by a number specifying the repetition period in days.
.PP
For example, suppose you get paid every second Wednesday, and your
last payday was Wednesday, 28 October, 1992.  You can use:
.PP
.nf
	REM 28 Oct 1992 *14 MSG Payday
.fi
.PP
This issues the reminder every 14 days, starting from the calculated trigger
date.  You can use \fIdelta\fR and \fIback\fR with \fIrepeat.\fR  Note,
however, that the \fIback\fR is used only to compute the initial
trigger date; thereafter, the reminder repeats with the specified
period.  Similarly, if you specify a weekday, it is used only to calculate
the initial date, and does not affect the repetition period.
.PP
.B SCANFROM
.PP
The \fBSCANFROM\fR keyword is for advanced \fBRemind\fR programmers
only, and will be explained in the section "Details about Trigger Computation"
near the end of this manual.  Note that \fBSCANFROM\fR is available only
in versions of \fBRemind\fR from 03.00.04 up.
.PP
.B PRIORITY
.PP
The \fBPRIORITY\fR keyword must be followed by a number from 0 to 9999.
It is used in calendar mode and when sorting reminders.  If two reminders
have the same trigger date and time, then they are sorted by priority.
If the \fBPRIORITY\fR keyword is not supplied, a default priority of 5000
is used.  (This default can be changed by adjusting the system variable
\fB$DefaultPrio\fR.  See the section "System Variables" for more
information.)
.PP
.B EXPIRY DATES
.PP
Some reminders should be issued periodically for a certain time, but then
expire.  For example, suppose you have a class every Friday, and that your
last class is on 11 December 1992.  You can use:
.PP
.nf
	REM Fri UNTIL 11 Dec 1992 MSG Class today.
.fi
.PP
Another example:  Suppose you have jury duty from 30 November 1992 until
4 December 1992.  The following reminder will issue the message every day
of your jury duty, as well as 2 days ahead of time:
.PP
.nf
	REM 30 Nov 1992 *1 +2 UNTIL 4 Dec 1992 MSG Jury duty
.fi
.PP
Note that the \fIrepeat\fR of *1 is necessary; without it, the reminder
would be issued only on 30 November (and the two days preceding.)
.PP
.B THE ONCE KEYWORD
.PP
Sometimes, it is necessary to ensure that reminders are run only once
on a given day.  For example, if you have a reminder which makes a backup
of your files every Friday:
.PP
.nf
	REM Fri RUN do_backup
.fi
.PP
(Here, \fIdo_backup\fR is assumed to be a program or shell script which
does the work.)  If you run \fBRemind\fR from your .login script, for
example, and log in several times per day, the \fIdo_backup\fR program
will be run each time you log in.  If, however, you use the \fBONCE\fR
keyword in the reminder, the \fBRemind\fR checks the last access date of
the reminder script.  If it is the same as the current date, \fBRemind\fR
assumes that it has already been run, and will not issue reminders containing
the \fBONCE\fR keyword.
.PP
Note that if you view or edit your reminder script, the last access date
will be updated, and the \fBONCE\fR keyword will not operate properly.
If you start \fBRemind\fR with the \fB\-o\fR option, then the \fBONCE\fR
keyword will be ignored.
.PP
.B LOCALLY OMITTING WEEKDAYS
.PP
The \fBOMIT\fR portion of the \fBREM\fR command is used to "omit" certain
days when counting the \fIdelta\fR or \fIback\fR.  It is specified using
the keyword \fBOMIT\fR followed by a list of weekdays.  Its action is
best illustrated with examples:
.PP
.nf
	REM 1 +1 OMIT Sat Sun MSG Important Event
.fi
.PP
This reminder is normally triggered on the first of every month, as well
as the day preceding it.  However, if the first of the month falls on a
Sunday or Monday, then the reminder is triggered starting from the
previous Friday.  This is because the \fIdelta\fR of +1 does not count
Saturday or Sunday when it counts backwards from the trigger date to
determine how much advance warning to give.
.PP
Contrast this with the use of "++1" in the above command.  In this case,

www/php/remind.php
<?php

class Remind
{
    # For validating commands we send to popen
    function is_valid_day($d) {
	return (preg_match('/^\d+$/', $d) &&
	       $d >= 1 && $d <= 31;
    }

    function is_valid_month($m) {
	return
	    ($m == 'January') ||
	    ($m == 'February') ||
	    ($m == 'March') ||
	    ($m == 'April') ||
	    ($m == 'May') ||
	    ($m == 'June') ||
	    ($m == 'July') ||
	    ($m == 'August') ||
	    ($m == 'September') ||
	    ($m == 'October') ||
	    ($m == 'November') ||
	    ($m == 'December');
    }

    function is_valid_year($y) {
	return preg_match('/^\d\d\d\d$/', $y) &&
	    $y >= 1900;
    }


    function get_el(&$array, $i)
    {
	if (!array_key_exists($i, $array)) return null;
	return $array[$i];
    }

    function get_elem($array, $indexes)
    {
	foreach ($indexes as $i) {
	    if (!is_array($array)) return null;
	    if (!array_key_exists($i, $array)) return null;
	    $array = $array[$i];
	}
	return $array;
    }

    function munge_entry($day, &$results, &$specials, &$options, $str, &$e) {
	return htmlspecialchars($str);
    }

    function format_entry($day, &$results, &$specials, &$options, &$e) {
	$special = $this->get_el($e, 'special');
	$body = $this->get_el($e, 'body');

	if ($body === null) $body = '';
	if ($special === null || $special == '*') {
	    return $this->munge_entry($day, $results, $specials, $options, $body, $e);
	}
	if ($special == 'COLOR' || $special == 'COLOUR') {
	    if (preg_match('/^(\d+)\s+(\d+)\s+(\d+)\s+(.*)/', $body, $matches)) {
		return sprintf('<span style="color: #%02x%02x%02x">%s</span>',
			       $matches[1] % 255,
			       $matches[2] % 255,
			       $matches[3] % 255,
			       $this->munge_entry($day, $results, $specials, $options, $matches[4], $e));
	    }
	    return 'Bad COLOR spec: ' . htmlspecialchars($body);
	}

	# HTML is passed through un-munged.
	if ($special == 'HTML') return $body;

	# Ignore unknown specials
	return '';
    }

    function format_entries($day, &$results, &$specials, &$options, &$entries) {
	$html = '';
	foreach ($entries as $e) {
	    $html .= '<div class="rem-entry">' . $this->format_entry($day, $results, $specials, $options, $e) . '</div>';
	}
	return $html;
    }

    function do_one_day($day, &$results, &$specials, &$options) {
	$class = $this->get_elem($specials, array('HTMLCLASS', $day, 0, 'body'));
	$shade = $this->get_elem($specials, array('SHADE', $day, 0, 'body'));
	$moon  = $this->get_elem($specials, array('MOON', $day, 0, 'body'));

	if ($class === null) $class = 'rem-cell';
	$bg = '';
	if ($shade !== null) {
	    if (preg_match('/(\d+)\s+(\d+)\s+(\d+)/', $shade, $matches)) {
		if ($matches[1] <= 255 && $matches[2] <= 255 && $matches[3] <= 255) {
		    $bg = sprintf(' style="background: #%02x%02x%02x"',
				  $matches[1], $matches[2], $matches[3]);
		}
	    }
	}
	$html = "<td class=\"$class\"$bg>";

	$week = $this->get_elem($specials, array('WEEK', $day, 0, 'body'));
	if ($week === null) {
	    $week = '';
	} else {
	    $week = ' ' . $week;
	}

	$moon_html = '';
	if ($moon !== null) {
	    $phase = -1;
	    if (preg_match('/(\d+)\s+(\S+)\s+(\S+)\s+(.*)$/', $moon, $matches)) {
		$phase = $matches[1];
		$moonsize = $matches[2];
		$fontsize = $matches[3];
		$msg = $matches[4];
	    } elseif (preg_match('/(\d+)/', $moon, $matches)) {
		$phase = $matches[1];
		$msg = '';
	    }
	    if ($phase >= 0) {
		if ($phase == 0) {
		    $img = 'newmoon.png';
		    $title = 'New Moon';
		    $alt = 'new';
		} elseif ($phase == 1) {
		    $img = 'firstquarter.png';
		    $title = 'First Quarter';
		    $alt = '1st';
		} elseif ($phase == 2) {
		    $img = 'fullmoon.png';
		    $alt = 'full';
		    $title = 'Full Moon';
		} else {
		    $img = 'lastquarter.png';
		    $alt = 'last';
		    $title = 'Last Quarter';
		}
		$base = rtrim($this->get_el($options, 'imgbase'), '/');
		if ($base !== null) {
		    $img = $base . '/' . $img;
		}
		$moon_html = '<div class="rem-moon">' . "<img width=\"16\" height=\"16\" alt=\"$alt\" title=\"$title\" src=\"$img\">" . htmlspecialchars($msg) . '</div>';
	    }
	}

	# Day number
	$html .= $moon_html . '<div class="rem-daynumber">' . $day . $week . '</div>';

	# And the entries
	$entries = $this->get_elem($results, array('entries', $day));
	if (is_array($entries) && count($entries) > 0) {
	    $html .= '<div class="rem-entries">';
	    $html .= $this->format_entries($day, $results, $specials, $options, $entries);
	    $html .= '</div>';
	}
	$html .= "</td>\n";
	return $html;
    }

    function generate_html(&$results, &$specials, &$options)
    {
	$monday_first = $results['monday_flag'];
	$first_col = $results['first_day'];
	if ($monday_first) {
	    $first_col--;
	    if ($first_col < 0) $first_col = 6;
	}

	$last_col = ($first_col + $results['days_in_mon'] -1) % 7;

	$html = '<table class="rem-cal"><caption class="rem-cal-caption">' .
	    htmlspecialchars($results['month']) . ' ' . htmlspecialchars($results['year']) .
	    "</caption>\n";

	$html .= '<tr class="rem-cal-hdr-row">';
	if (!$monday_first) $html    .= '<th class="rem-cal-hdr">' . htmlspecialchars($results['day_names'][0])  . '</th>';
	for ($i=1; $i<7; $i++) $html .= '<th class="rem-cal-hdr">' . htmlspecialchars($results['day_names'][$i]) . '</th>';
	if ($monday_first) $html     .= '<th class="rem-cal-hdr">' . htmlspecialchars($results['day_names'][0])  . '</th>';
	$html .= "</tr>\n";

	# Do the leading empty columns
	for ($col=0; $col < $first_col; $col++) {
	    if ($col == 0) $html .= '<tr class="rem-cal-body-row">';
	    $html .= '<td class="rem-empty">&nbsp;</td>';
	}

	for ($day=1; $day <= $results['days_in_mon']; $day++) {
	    if ($col == 0) $html .= '<tr class="rem-cal-body-row">';
	    $col++;
	    $html .= $this->do_one_day($day, $results, $specials, $options);
	    if ($col == 7) {
		$html .= "</tr>\n";
		$col = 0;
	    }
	}
	if ($col) {
	    while ($col++ < 7) {
		$html .= '<td class="rem-empty">&nbsp;</td>';
	    }
	}
	$html .= "</tr>\n";

	$html .= "</table>\n";
	return $html;
    }
    function parse_remind_output ($fp)
    {
	while(1) {
	    $line = fgets($fp);
	    if ($line === false) break;
	    $line = trim($line);
	    if ($line == '# rem2ps begin') break;
	}
	if ($line === false) {
	    return array('success' => 0,
			 'error' => 'Could not find any Rem2PS data');
	}

	$line = fgets($fp);
	if ($line === false) {
	    return array('success' => 0,
			 'error' => 'Unexpected end-of-file');
	}

	$line = trim($line);
	list($month, $year, $days_in_mon, $first_day, $monday_flag) = explode(' ', $line);
	$retval = array('month' => $month,
			'year'  => $year,
			'days_in_mon' => $days_in_mon,
			'first_day' => $first_day,
			'monday_flag' => $monday_flag);

	$line = fgets($fp);
	if ($line === false) {
	    return array('success' => 0,
			 'error' => 'Unexpected end-of-file');
	}

	$line = trim($line);
	$retval['day_names'] = explode(' ', $line);

	$line = fgets($fp);
	if ($line === false) {
	    return array('success' => 0,
			 'error' => 'Unexpected end-of-file');
	}
	$line = trim($line);

	list($m, $n) = explode(' ', $line);
	$retval['prev'] = array('month' => $m, 'days' => $n);

	$line = fgets($fp);
	if ($line === false) {
	    return array('success' => 0,
			 'error' => 'Unexpected end-of-file');
	}
	$line = trim($line);

	list($m, $n) = explode(' ', $line);
	$retval['next'] = array('month' => $m, 'days' => $n);

	$line_info = 0;

	$entries = array();
	$specials = array();
	while (1) {
	    $line = fgets($fp);
	    if ($line === false) break;
	    $line = trim($line);
	    if ($line == '# rem2ps end') break;
	    if (strpos($line, '# fileinfo ') === 0) {
		list($lno, $fname) = explode(' ', substr($line, 11), 2);
		$lineinfo = array('file' => $fname, 'line' => $lno);
		continue;
	    }
	    list($date, $special, $tags, $duration, $time, $body) = explode(' ', $line, 6);
	    list($y, $m, $d) = explode('/', $date);
	    $d = preg_replace('/^0(.)/', '$1', $d);
	    $m = preg_replace('/^0(.)/', '$1', $m);
	    $entry = array('day' => $d,
			   'month' => $m,
			   'year' => $y,
			   'special' => $special,
			   'tags' => $tags,
			   'duration' => $duration,
			   'time' => $time,
			   'body' => $body);
	    if (is_array($lineinfo)) {
		$entry['line'] = $lineinfo['line'];
		$entry['file'] = $lineinfo['file'];
		$lineinfo = 0;
	    }
	    if ($special != '*' && $special != 'COLOR' && $special != 'COLOUR' && $special != 'HTML') {
		if (!array_key_exists($special, $specials)) {
		    $specials[$special] = array();
		}
		if (!array_key_exists($d, $specials[$special])) {
		    $specials[$special][$d] = array();
		}
		$specials[$special][$d][] = $entry;
	    } else {
		if (!array_key_exists($d, $entries)) {
		    $entries[$d] = array();
		}
		$entries[$d][] = $entry;
	    }

	}
	$retval['entries'] = $entries;
	return array('success' => 1, 'results' => $retval, 'specials' => $specials);
    }
}

$fp = popen('rem -p -l', 'r');
$r = new Remind;
$ans = $r->parse_remind_output($fp);
pclose($fp);
#print_r($ans);
$options = array();
print $r->generate_html($ans['results'], $ans['specials'], $options);

?>
